hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/opt/OptUtils.java

```diff
@@ -172,31 +173,35 @@ public static boolean isPhysical(RelNode rel) {
     /**
      * Get possible physical rels from the given subset. Every returned input is guaranteed to have a unique trait set.
      *
-     * @param subset Subset.
+     * @param input Subset.
      * @return Physical rels.
      */
-    public static Collection<RelNode> getPhysicalRelsFromSubset(RelNode subset) {
-        if (subset instanceof RelSubset) {
-            RelSubset subset0 = (RelSubset) subset;
+    public static Collection<RelNode> getPhysicalRelsFromSubset(RelNode input) {
 
-            Set<RelTraitSet> traitSets = new HashSet<>();
+        RelNode convertedInput = convert(input, input.getTraitSet());
 
-            Set<RelNode> res = Collections.newSetFromMap(new IdentityHashMap<>());
+        if (!(convertedInput instanceof RelSubset)) {
+            return Collections.emptyList();
+        }
+
+        RelSubset logicalSubset = (RelSubset) convertedInput;
+        RelSubset physicalSubset = HazelcastRelSubsetUtil.getPhysicalSubSet(logicalSubset);
+
+        Set<RelTraitSet> traitSets = new HashSet<>();
 
-            for (RelNode rel : subset0.getRelList()) {
-                if (!isPhysical(rel)) {
-                    continue;
-                }
+        Set<RelNode> res = Collections.newSetFromMap(new IdentityHashMap<>());
 
-                if (traitSets.add(rel.getTraitSet())) {
-                    res.add(convert(subset, rel.getTraitSet()));
-                }
+        for (RelNode rel : physicalSubset.getRelList()) {
+            if (!isPhysical(rel)) {
+                continue;
             }
 
-            return res;
-        } else {
-            return Collections.emptyList();
+            if (traitSets.add(rel.getTraitSet())) {
+                res.add(convert(physicalSubset, rel.getTraitSet()));
```

Comment:
Here we convert a physical subset to another physical rel traits. I do not know the exact consequences of it, but perhaps it would be better to do `convert(rel, rel.getTraitSet())`. In any case, let's double-check that this `convert` call doesn't lead to the creation of new converters.

Reply:
Addressed in acd0028